### ISSUE TITLE
Generic Bridge to backend

### ISSUE THREAD

**Issue by aslesarenko** [scalan/scalan#97]
Currently each node is transfered to LMS using code like this

```scala
case ArraySort(arg, o) =>
  arg.elem match {
    case (el: ArrayElem[_]) =>
      createManifest(el.eItem) match {
        case (mA: Manifest[a]) =>
          val arg_ = m.symMirror[Array[a]](arg)
          val exp = lms.arraySort[a](arg_)(mA)
          m.addSym(sym, exp)
      }
  }
```
This is clearly complete boilerplate.

##### Considerations for implementation:
* load backend object dynamically (Dependency Injection style)
* transfer data to backend using either Reflection of some other generic interface (this interface can be called for example BridgeTarget)
* concrete class that implements this interface can be unknown and the time of compilation

**Comment by alexeyr** [comment]
Applies to createManifest as well.

**Comment by alexeyr** [comment]
This will likely have to be modified to reuse `Proxy`'s element extraction.

**Comment by alexeyr** [comment]
Current implementation seems to work well, now checking what precisely we can replace. See https://github.com/scalan/scalan/tree/generic-bridge.

**Comment by aslesarenko** [comment]
:+1: for second subitem
I suggest to focus on the first subitem

**Comment by alexeyr** [comment]
First subitem: we already can compile and publish backends separately. They would just need to be overridden in the compiler. We can actually move backend to a constructor argument (with a default value) and it could be managed with constructor injection if required.

**Comment by alexeyr** [comment]
I could also move backend (but not bridge) into a separate sub-project which doesn't depend on the rest of Scalan, as discussed before, to show this independence. Should this be done now?

**Comment by alexeyr** [comment]
Alternate solution to first subitem is not to mix Bridge into Compiler but instead pass it. I think it isn't useful enough to do currently.

**Comment by aslesarenko** [comment]
The first and the last subitems are closely related, IMO.
**The key point of the first** is the ability to load dynamically (and even reload at runtime something different), compiled, and what is more important, deployed separately.
**The key point of the last** is to have *well defined contract* which is implemented by all `Backend`s.
Together they will allow to implement Bridge as generic function (*once and for all*). 
Implemented this way, Bridge is ok to be in scalan.core and mixed-in in Compiler cake.

**Comment by alexeyr** [comment]
> is the ability to load dynamically (and even reload at runtime something different), compiled, and what is more important, deployed separately

Compiling and deploying separately can be done by moving into a separate subproject as above.

Loading dynamically by passing it at construction (again, DI can be used for this). But reloading at runtime is problematic (would mean it isn't a stable value and can't be used in types). Of course, we can freely create multiple compilers with different backends.

> The key point of the last is to have well defined contract which is implemented by all Backends.
Together they will allow to implement Bridge as generic function (once and for all).
Implemented this way, Bridge is ok to be in scalan.core and mixed-in in Compiler cake.

It _could_ be pulled out into scalan-core along with a Backend but removing LMS dependency from Backend contract will require duplicating a lot of method signatures, and with no benefit until we actually have non-LMS compilers _and_ decide to use Bridge for them. I don't think it should be done now.

**Comment by aslesarenko** [comment]
Let's reconsider this after #109 is implemented.

**Comment by alexeyr** [comment]
Move to a separate subproject can be done now but in a separate pull request to simplify merging (e.g. @yzotov has some changes here).

**Comment by aslesarenko** [comment]
Ok.
